aptos-move/framework/src/natives/cryptography/bls12381.rs

```diff
@@ -0,0 +1,696 @@
+* native fun native_bls12381_verify_aggregate_signature
+*
+*   gas cost: base_cost + per_pubkey_deserialize_cost * min(num_validatable_pubkeys + 1, num_pubkeys)
+*                       +? ( per_sig_deserialize_cost * min(num_viable_sigs + 1, num_sigs)
+*                            +? ( per_pairing_cost + per_msg_hashing_cost ) * num_msgs
+*                                 + per_byte_hashing_cost * total_msg_bytes )
+*
+* where:
+*    +? indicates the expression stops evaluating there if the previous gas-charging step failed,
+*    num_pubkeys is the # of public keys given as input,
+*    num_validatable_pubkeys is the # of public keys that deserialize successfully (i.e., "validatable"),
+*    num_sigs is the # of signatures given as input,
+*    num_viable_sigs is the # of signatures that deserialize successfully (i.e., "viable"),
+*    total_msg_bytes is the cumulative size in bytes of all messages.
+*
+* NOTE(ValidatablePK): We refer to the public keys that deserialize correctly as "validatable"
+* above, since successful deserialization is not a sufficient criteria for the "validatability"
+* of a PK: e.g., the PK could still be susceptible to small-subgroup attacks or rogue-key attacks.
+*
+* NOTE(ViableSig): We refer to the signatures that deserialize correctly as "viable" above, since
+* successful deserialization is not a sufficient criteria for the "viability" of a signature: e.g.,
+* the signature might not verify under the desired (message, public key) pair, or the signature
+* could lie in a small-order subgroup.
+
+* NOTE: If all PKs deserialize, then per_pubkey_deserialize_cost is charged num_pubkeys times.
+* Otherwise, if only num_validatable_pubkeys deserialize correctly, an extra per_pubkey_deserialize_cost
+* must be charged for the failed deserialization. We proceed similarly for per_sig_deserialize_cost.
+**************************************************************************************************/
+pub fn native_bls12381_verify_aggregate_signature(
+    gas_params: &GasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 3);
+
+    let mut cost = gas_params.base_cost;
+
+    // Parses a Vec<Vec<u8>> of all messages
+    let messages = pop_vec_arg!(arguments, Vec<u8>);
+    // Parses a Vec<Vec<u8>> of all serialized public keys
+    let pks_serialized = pop_vec_of_vec_u8(&mut arguments)?;
+    let num_pks = pks_serialized.len();
+
+    // Parses the signature as a Vec<u8>
+    let aggsig_bytes = pop_arg!(arguments, Vec<u8>);
+
+    // Number of messages must match number of public keys
+    if pks_serialized.len() != messages.len() {
+        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
+    }
+
+    let pks = bls12381_deserialize_pks_with_gas(pks_serialized, &mut cost, gas_params);
```

Comment:
my recommendation is to implement all these functions for GasParameters, so you'd have
```
impl GasParas {
  fn deserialize_pks(&self, pks_serialized: vector<u8>, cost: &mut GasCost) -> Result<Vec<PublicKey>> {
   
  }
}
```

Reply:
Fixed!